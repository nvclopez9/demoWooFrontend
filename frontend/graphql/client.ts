import { GraphQLClient } from 'graphql-request';
import deepmerge from 'deepmerge';

import {
  RootQueryToProductConnectionWhereArgs,
  Product,
  ProductCategory,
  PaColor,
  getSdk,
  RootQueryToProductCategoryConnectionWhereArgs,
  RootQueryToPaColorConnectionWhereArgs,
  ProductIdTypeEnum,
} from './generated';

let client: GraphQLClient;
export function getClient() {
  const endpoint = process.env.GRAPHQL_ENDPOINT
  if (!endpoint) {
    throw new Error('GRAPHQL_ENDPOINT is not defined')
  }

  if (!client) {
    client = new GraphQLClient(endpoint);
  }

  return client;
}

export function getClientWithSdk() {
  return getSdk(getClient());
}

const initialConnectionResults = {
  pageInfo: {
    hasNextPage: true,
    endCursor: null,
  },
  edges: [],
  nodes: [],
};

export async function fetchProducts(
  pageSize: number, 
  pageLimit = 0,
  where?: RootQueryToProductConnectionWhereArgs) {
  try {
    const client = getClientWithSdk();
    let data = { products: initialConnectionResults }
    let after = '';
    let count = 0
    while(data.products.pageInfo.hasNextPage && (pageLimit === 0 || count < pageLimit)) {
      const next = await client.GetProducts({
        first: pageSize,
        after,
        where,
      });

      data = deepmerge(data, next);
      after = next.products?.pageInfo.endCursor || '';
      count++;
    }

    return (data.products.nodes) as Product[];
  } catch (err) {
    console.error(err || 'Failed to fetch product listing!!!');
  }
}

export async function fetchCategories(
  pageSize: number,
  pageLimit = 0,
  where?: RootQueryToProductCategoryConnectionWhereArgs
) {
  try {
    const client = getClientWithSdk();
    let data = { productCategories: initialConnectionResults }
    let after = '';
    let count = 0
    while(data.productCategories.pageInfo.hasNextPage && (pageLimit === 0 || count < pageLimit)) {
      const next = await client.GetShopCategories({
        first: pageSize,
        after,
        where,
      });

      data = deepmerge(data, next);
      after = next.productCategories?.pageInfo.endCursor || '';
      count++;
    }

    return (data.productCategories.nodes) as ProductCategory[];

  } catch (err) {
    console.error(err || 'Failed to fetch product categories!!!');
  }
}

export async function fetchColors(
  pageSize: number,
  pageLimit = 0,
  where?: RootQueryToPaColorConnectionWhereArgs
) {
  try {
    const client = getClientWithSdk();
    let data = { allPaColor: initialConnectionResults }
    let after = '';
    let count = 0
    while(data.allPaColor.pageInfo.hasNextPage && (pageLimit === 0 || count < pageLimit)) {
      const next = await client.GetShopColors({
        first: pageSize,
        after,
        where
      });

      data = deepmerge(data, next);
      after = next.allPaColor?.pageInfo.endCursor || '';
      count++;
    }

    return (data.allPaColor.nodes) as PaColor[];
  } catch (err) {
    console.error(err || 'Failed to fetch product color attributes!!!');
  }
}

export async function fetchProductBy(slug: string, idType: ProductIdTypeEnum) {
  try {
    const client = getClientWithSdk();
    const data = await client.GetProduct({
      id: slug,
      idType: idType,
    });

    if (!data.product) {
      throw new Error('Product not found!!!');
    }

    return data.product as Product;
  } catch (err) {
    console.error(err || 'Failed to fetch product data!!!');
  }
}