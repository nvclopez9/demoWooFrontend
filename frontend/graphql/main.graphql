fragment MenuItemContent on MenuItem {
  id
  uri
  title
  label
  cssClasses
}

fragment MenuItemRecursive on MenuItem {
  ...MenuItemContent
  childItems {
    nodes {
      ...MenuItemContent
    }
  }
}

fragment MenuContent on Menu {
  id
  name
  locations
  slug
  menuItems(first: 20, where: {parentId: 0}) {
    nodes {
      ...MenuItemRecursive
    }
  }
}

fragment CustomerContent on Customer {
  id
  sessionToken
  firstName
  shipping {
    postcode
    state
    city
    country
  }
}

fragment ProductContentSlice on Product {
  id
  databaseId
  name
  slug
  type
  image {
    id
    sourceUrl(size:WOOCOMMERCE_THUMBNAIL)
    altText
  }
  ... on SimpleProduct {
    price
    regularPrice
    soldIndividually
  }
  ... on VariableProduct {
    price
    regularPrice
    soldIndividually
  }
}

fragment ProductVariationContentSlice on ProductVariation {
  id
  databaseId
  name
  slug
  image {
    id
    sourceUrl(size:WOOCOMMERCE_THUMBNAIL)
    altText
  }
  price
  regularPrice
}

fragment ProductContentSmall on Product {
  id
  databaseId
  slug
  name
  type
  shortDescription(format: RAW)
  image {
    id
    sourceUrl(size:WOOCOMMERCE_THUMBNAIL)
    altText
  }
  productCategories(first: 20) {
    nodes {
      id
      slug
      name
    }
  }
  productTags(first: 20) {
    nodes {
      id
      slug
      name
    }
  }
  allPaColor(first: 100) {
    nodes {
      id
      slug
      name
    }
  }
  ... on SimpleProduct {
    onSale
    stockStatus
    price
    rawPrice: price(format: RAW)
    regularPrice
    salePrice
    soldIndividually
  }
  ... on VariableProduct {
    onSale
    stockStatus
    price
    rawPrice: price(format: RAW)
    regularPrice
    salePrice
    soldIndividually
  }
}

fragment ProductContentFull on Product {
  id
  databaseId
  slug
  name
  type
  description
  shortDescription(format: RAW)
  image {
    id
    sourceUrl
    altText
  }
  galleryImages {
    nodes {
      id
      sourceUrl(size:WOOCOMMERCE_THUMBNAIL)
      altText
    }
  }
  productTags(first: 20) {
    nodes {
      id
      slug
      name
    }
  }
  attributes {
    nodes {
      id
      attributeId
      ... on LocalProductAttribute {
        name
        label
        options
        variation
      }
      ... on GlobalProductAttribute {
        name
        label
        options
        variation
        slug
        terms(first: 100) {
          nodes {
            id
            name
            slug
          }
        }
      }
    }
  }
  ... on SimpleProduct {
    onSale
    stockStatus
    price
    rawPrice: price(format: RAW)
    regularPrice
    salePrice
    stockStatus
    stockQuantity
    soldIndividually
    defaultAttributes(first: 100) {
      nodes {
        id
        attributeId
        name
        value
        label
      }
    }
  }
  ... on VariableProduct {
    onSale
    price
    rawPrice: price(format: RAW)
    regularPrice
    salePrice
    stockStatus
    stockQuantity
    soldIndividually
    defaultAttributes(first: 100) {
      nodes {
        id
        attributeId
        label
        name
        value
      }
    }
    variations(first: 50) {
      nodes {
        id
        databaseId
        name
        price
        rawPrice: price(format: RAW)
        regularPrice
        salePrice
        onSale
        attributes {
          nodes {
            name
            label
            value
          }
        }
        image {
          id
          sourceUrl
          altText
        }
      }
    }
  }
}

fragment VariationContent on ProductVariation {
  id
  name
  slug
  price
  regularPrice
  salePrice
  stockStatus
  stockQuantity
  onSale
  image {
    id
    sourceUrl
    altText
  }
}

fragment CartItemContent on CartItem {
  key
  product {
    node {
      ...ProductContentSlice
    }
  }
  variation {
    attributes {
      id
      label
      name
      value
    }
    node {
      ...ProductVariationContentSlice
    }
  }
  quantity
  total
  subtotal
  subtotalTax
  extraData {
    key
    value
  }
}

fragment CartContent on Cart {
  contents(first: 100) {
    itemCount
    nodes {
      ...CartItemContent
    }
  }
  appliedCoupons {
    code
    discountAmount
    discountTax
  }
  needsShippingAddress
  availableShippingMethods {
    packageDetails
    supportsShippingCalculator
    rates {
      id
      instanceId
      methodId
      label
      cost
    }
  }
  subtotal
  subtotalTax
  shippingTax
  shippingTotal
  total
  totalTax
  feeTax
  feeTotal
  discountTax
  discountTotal
}

fragment CustomerFields on Customer {
    id
    databaseId
    firstName
    lastName
    displayName
    email
    sessionToken
    metaData {
      key
      value
    }
}

query GetTopNav {
  menu(id: "primary", idType: LOCATION) {
    ...MenuContent
  }
}

query GetProducts($first: Int, $after: String, $where: RootQueryToProductConnectionWhereArgs) {
  products(first: $first, after: $after, where: $where) {
    pageInfo {
      endCursor
      hasNextPage
    }
    edges {
      cursor
      node {
        ...ProductContentSmall
      }
    }
    nodes {
      ...ProductContentSmall
    }
  }
}

query GetProduct($id: ID!, $idType: ProductIdTypeEnum) {
  product(id: $id, idType: $idType) {
    ...ProductContentFull
  }
}

query GetShopCategories($first: Int, $after: String, $where: RootQueryToProductCategoryConnectionWhereArgs) {
  productCategories(first: $first, after: $after, where: $where) {
    pageInfo {
      hasNextPage
      endCursor
    }
    edges {
      node {
        id
        name
        slug
      }
    }
    nodes {
      id
      name
      slug
    }
  }
}

query GetShopTags($first: Int, $after: String, $where: RootQueryToProductTagConnectionWhereArgs) {
  productTags(first: $first, after: $after, where: $where) {
    pageInfo {
      hasNextPage
      endCursor
    }
    edges {
      node {
        id
        name
        slug
      }
    }
    nodes {
      id
      name
      slug
    }
  }
}

query GetShopColors($first: Int, $after: String, $where: RootQueryToPaColorConnectionWhereArgs) {
  allPaColor(first: $first, after: $after, where: $where) {
    pageInfo {
      hasNextPage
      endCursor
    }
    edges {
      node {
        id
        name
        slug
      }
    }
    nodes {
      id
      name
      slug
    }
  }
}

query GetSession {
  cart {
    ...CartContent
  }
  customer {
    ...CustomerContent
  }
}

mutation AddToCart(
  $productId: Int!,
  $variationId: Int,
  $quantity: Int,
  $variation: [ProductAttributeInput],
  $extraData: String
) {
  addToCart(
    input: {
      productId: $productId,
      variationId: $variationId,
      quantity: $quantity,
      variation: $variation,
      extraData: $extraData
    }
  ) {
    cart {
      ...CartContent
    }
    cartItem {
      ...CartItemContent
    }
  }
}

mutation UpdateCartItemQuantities($items: [CartItemQuantityInput]) {
  updateItemQuantities(input: {items: $items}) {
    cart {
      ...CartContent
    }
    items {
      ...CartItemContent
    }
  }
}

mutation RemoveItemsFromCart($keys: [ID], $all: Boolean) {
  removeItemsFromCart(input: {keys: $keys, all: $all}) {
    cart {
      ...CartContent
    }
    cartItems {
      ...CartItemContent
    }
  }
}

mutation Login($username: String!, $password: String!) {
  login(input: { username: $username, password: $password }) {
    authToken
    refreshToken
    customer {
      ...CustomerFields
    }
  }
}

mutation RefreshAuthToken($refreshToken: String!) {
  refreshJwtAuthToken(input: { jwtRefreshToken: $refreshToken }) {
    authToken
  }
}

mutation UpdateSession($input: UpdateSessionInput!) {
  updateSession(input: $input) {
    session {
      id
      key
      value
    }
  }
}