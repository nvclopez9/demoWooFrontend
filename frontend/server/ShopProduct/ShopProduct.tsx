import { Product, SimpleProduct, VariationAttribute } from '@woographql/graphql';
import { CartOptions } from '@woographql/server/CartOptions';
import { ProductImage } from '@woographql/client/ProductImage';
import {
  Tabs,
  TabsContent,
  TabsList,
  TabsTrigger,
} from '@woographql/ui/tabs';

export interface ShopProductProps {
  product: Product;
}

export function ShopProduct(props: ShopProductProps) {
  const {
    product,
  } = props;

  const attributes: Record<string, string[]> = (product as SimpleProduct).defaultAttributes?.nodes?.reduce(
    (attributesList, attribute) => {
      const {
        value,
        label
      } = attribute as VariationAttribute;

      const currentAttributes = attributesList[label as string] || [];
      return {
        ...attributesList,
        [label as string]: [...currentAttributes, value as string],
      };
    },
    {} as Record<string, string[]>
  ) || {};

  return (
    <div className="w-full flex flex-wrap gap-4 max-w-screen-lg mx-auto mb-36 lg:mb-0">
      <ProductImage product={product} />
      <div className="basis-full md:basis-1/2 pt-4 px-4 flex flex-col">
        <h1 className="font-serif text-2xl font-bold mb-2">{product.name}</h1>
        <p className="text-lg font-bold mb-2">{product.shortDescription}</p>
        <CartOptions product={product} />
      </div>
      <Tabs defaultValue="description" className="w-full px-4">
        <TabsList className="w-full md:w-auto">
          <TabsTrigger value="description">Description</TabsTrigger>
          <TabsTrigger value="attributes">Attributes</TabsTrigger>
          <TabsTrigger value="reviews">Reviews</TabsTrigger>
        </TabsList>
        <TabsContent value="description">
          <div dangerouslySetInnerHTML={{ __html: product.description as string }} />
        </TabsContent>
        <TabsContent value="attributes">
          <ul>
            {Object.entries(attributes).map(([label, values]) => (
              <li key={label}>
                <p><span className="font-serif font-medium">{label}:</span> {values.join(', ')}</p>
              </li>
            ))}
          </ul>
        </TabsContent>
        <TabsContent value="reviews">
          <p>Reviews</p>
        </TabsContent>
      </Tabs>
    </div>
  );
}
