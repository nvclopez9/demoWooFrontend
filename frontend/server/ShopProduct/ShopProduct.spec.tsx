// ShopProduct.test.tsx
import { SimpleProduct, VariationAttribute } from '@woographql/graphql';
import { render, screen, product } from '@woographql/testing';
import { ShopProduct } from './ShopProduct';

describe('ShopProduct component', () => {
  it('renders correctly', () => {
    render(<ShopProduct product={product} />);

    // Check that the product name, short description, and description are rendered
    expect(screen.getByText(product.name as string)).toBeInTheDocument();
    expect(screen.getByText(product.shortDescription as string)).toBeInTheDocument();
    expect(screen.getByText(product.description as string)).toBeInTheDocument();

    // Check that the attribute names and values are rendered
    (product as SimpleProduct).defaultAttributes?.nodes.forEach((attribute: VariationAttribute) => {
      expect(screen.getByText((attribute.name as string) + ':')).toBeInTheDocument();
      expect(screen.getByText(attribute.value as string)).toBeInTheDocument();
    });

    // Check that the Description, Attributes, and Reviews tabs are rendered
    expect(screen.getByText('Description')).toBeInTheDocument();
    expect(screen.getByText('Attributes')).toBeInTheDocument();
    expect(screen.getByText('Reviews')).toBeInTheDocument();
  });

  it('should match snapshot', () => {
    const { baseElement } = render(<ShopProduct product={product} />);
    expect(baseElement).toMatchSnapshot();
  });
});
