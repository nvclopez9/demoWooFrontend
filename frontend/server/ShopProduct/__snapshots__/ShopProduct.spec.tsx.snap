// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShopProduct component should match snapshot 1`] = `
<body>
  <div>
    <div
      class="w-full flex flex-wrap gap-4 max-w-screen-lg mx-auto mb-36 lg:mb-0"
    >
      <div
        class="basis-full md:basis-1/2 pt-4 px-4 flex flex-col"
      >
        <h1
          class="font-serif text-2xl font-bold mb-2"
        >
          T Shirt
        </h1>
        <p
          class="text-lg font-bold mb-2"
        >
          Short description 1
        </p>
        <div
          class="fixed inset-x-0 mx-auto p-4 lg:relative lg:inset-auto lg:mx-0 lg:p-0 bg-white bottom-0 z-30 w-screen lg:bg-inherit lg:bottom-auto lg:z-auto lg:w-auto"
        />
      </div>
      <div
        class="w-full px-4"
        data-orientation="horizontal"
        dir="ltr"
      >
        <div
          aria-orientation="horizontal"
          class="inline-flex h-9 items-center justify-center rounded-lg bg-muted p-1 text-muted-foreground w-full md:w-auto"
          data-orientation="horizontal"
          role="tablist"
          style="outline: none;"
          tabindex="0"
        >
          <button
            aria-controls="radix-:r4:-content-description"
            aria-selected="true"
            class="inline-flex items-center justify-center whitespace-nowrap rounded-md px-3 py-1 text-sm font-medium ring-offset-background transition-all focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 data-[state=active]:bg-background data-[state=active]:text-foreground data-[state=active]:shadow"
            data-orientation="horizontal"
            data-radix-collection-item=""
            data-state="active"
            id="radix-:r4:-trigger-description"
            role="tab"
            tabindex="-1"
            type="button"
          >
            Description
          </button>
          <button
            aria-controls="radix-:r4:-content-attributes"
            aria-selected="false"
            class="inline-flex items-center justify-center whitespace-nowrap rounded-md px-3 py-1 text-sm font-medium ring-offset-background transition-all focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 data-[state=active]:bg-background data-[state=active]:text-foreground data-[state=active]:shadow"
            data-orientation="horizontal"
            data-radix-collection-item=""
            data-state="inactive"
            id="radix-:r4:-trigger-attributes"
            role="tab"
            tabindex="-1"
            type="button"
          >
            Attributes
          </button>
          <button
            aria-controls="radix-:r4:-content-reviews"
            aria-selected="false"
            class="inline-flex items-center justify-center whitespace-nowrap rounded-md px-3 py-1 text-sm font-medium ring-offset-background transition-all focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 data-[state=active]:bg-background data-[state=active]:text-foreground data-[state=active]:shadow"
            data-orientation="horizontal"
            data-radix-collection-item=""
            data-state="inactive"
            id="radix-:r4:-trigger-reviews"
            role="tab"
            tabindex="-1"
            type="button"
          >
            Reviews
          </button>
        </div>
        <div
          aria-labelledby="radix-:r4:-trigger-description"
          class="mt-2 ring-offset-background focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2"
          data-orientation="horizontal"
          data-state="active"
          id="radix-:r4:-content-description"
          role="tabpanel"
          style="animation-duration: 0s;"
          tabindex="0"
        >
          <div>
            Description 1
          </div>
        </div>
        <div
          aria-labelledby="radix-:r4:-trigger-attributes"
          class="mt-2 ring-offset-background focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2"
          data-orientation="horizontal"
          data-state="inactive"
          hidden=""
          id="radix-:r4:-content-attributes"
          role="tabpanel"
          tabindex="0"
        />
        <div
          aria-labelledby="radix-:r4:-trigger-reviews"
          class="mt-2 ring-offset-background focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2"
          data-orientation="horizontal"
          data-state="inactive"
          hidden=""
          id="radix-:r4:-content-reviews"
          role="tabpanel"
          tabindex="0"
        />
      </div>
    </div>
  </div>
</body>
`;
